### PR TITLE
Fix #6315 - block certain data URL as top-level navigation

### LINKED ISSUE #6315: Disable certain 'data:' URLs types in address bar
https://bugzilla.mozilla.org/show_bug.cgi?id=1624167

Certain data URL types should not be allowed, see https://blog.mozilla.org/security/2017/11/27/blocking-top-level-navigations-data-urls-firefox-59/

The important bit from that link is to allow the following:
``` the following cases will be allowed:
    1) User explicitly entering/pasting “data:…” into the address bar
    2) Opening all plain text data files
    3) Opening “data:image/*” in top-level window, unless it’s “data:image/svg+xml”
    4) Opening “data:application/pdf” and “data:application/json”
    5) Downloading a data: URL, e.g. ‘save-link-as’ of “data:…”
```

To avoid security risks we are going to only allow 2,3, and 4. The others are too risky IMO. 
That said, if there is interest in attempting 1 and 5 in a secure manner, please ping me before attempting this, not to dissuade you from it, but there might be additional consideration to discuss.

The test page for this is https://www-archive.mozilla.org/quality/networking/testing/datatests.html

### PR DESCRIPTION
**Pull request by garvankeeley**
Please refer to https://github.com/mozilla-mobile/firefox-ios/issues/6315 for details on this PR.

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift [hunk 1]
```diff
@@ -431,10 +431,35 @@ extension BrowserViewController: WKNavigationDelegate {
...
+                return
+            }
+
+            // Allow certain application types
+            if url.hasPrefix("data:application/pdf") || url.hasPrefix("data:application/json") {
+                decisionHandler(.allow)
+                return
+            }
+
+            // Allow plan text types
```
Comment (nbhasin2) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift]: nit: typo plan -> **plain**

File: UITests/SecurityTests.swift [hunk 2]
```diff
@@ -120,9 +120,35 @@ class SecurityTests: KIFTestCase {
...
+        ["data-url-ok-1", "data-url-ok-2"].forEach { buttonName in
+            tester().tapWebViewElementWithAccessibilityLabel(buttonName)
+            tester().wait(forTimeInterval: 1)
+            let webView = tester().waitForView(withAccessibilityLabel: "Web content") as! WKWebView
+            XCTAssert(webView.url!.absoluteString.hasPrefix("data:")) // indicates page loaded ok
+            BrowserUtils.resetToAboutHome()
+            beforeEach()
+        }
+
+        // Check data url that is no allowed
```
Comment (nbhasin2) [inline UITests/SecurityTests.swift]: nit: no allowed -> **not allowed**

**Comment by Brandon-T** [comment]
@garvankeeley 

This PR potentially blocks some videos on vimeo that uses the `data:video/mp4` URI scheme.
Firefox desktop also allows `data:video/*`.

Tested on FF Desktop v75 (64-bit), MacOS Catalina.

**Comment by garvankeeley** [comment]
Thanks for letting me know @Brandon-T !! Will fix that.
BTW, that synthetic click blocker you figured out is really cool. Been meaning to look at that more closely.

**Pull request merged**